### PR TITLE
ARROW-13016: [C++][Compute] Support Null type in Sum/Mean aggregation

### PR DESCRIPTION
**Pull request by Crystrix**
The Sum/Mean of a Null type array is a Null type scalar.

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
One problem with supporting this for the "null" type, is that a sum of an empty array (empty to start with or after skipping all the nulls) can be defined as 0 (depending on the value of the `min_count` parameter), while for "null" type that can of course only be null.

**Comment by pitrou** [comment]
@jorisvandenbossche has a good point. The only reason to define these kernels is for consistency. But always returning null doesn't seem consistent with current behaviour of the kernels for numeric types.

@bkietz Any opinion?

**Comment by Crystrix** [comment]
Maybe we can keep a consistent behavior with numeric types. 
For example,
- Count: keep the current implementation (return 0 for `skip_nulls` option, otherwise return the count of the null elements)
- Sum: return 0 if `min_count` == 0, otherwise return a null scalar
- Mean: return a double scalar of null value
- MinMax: return a struct scalar of two null scalar children

**Comment by jorisvandenbossche** [comment]
`Count` is indeed certainly not a problem for "null" type (since it only counting the values, not calculating anything with them). And also for `MinMax` it seems this always returns null for empty or all-null numeric arrays, so there wouldn't be an inconsistency here. 

So it's mainly `Sum` that would give a type inconsistency. `Mean` might give less of a problem, because currently that kernel always returns a double regardless of the (numeric) input type.

File: cpp/src/arrow/compute/exec.cc [hunk 1]
```diff
@@ -961,6 +961,14 @@ class ScalarAggExecutor : public KernelExecutorImpl<ScalarAggregateKernel> {
     return outputs[0];
   }
 
+  Status CheckResultType(const Datum& out, const char* function_name) override {
+    if (strcmp(function_name, "sum") == 0 && output_descr_.type->Equals(null())) {
+      // Skip CheckResultType for sum of null type, as it may return various types
+      return Status::OK();
+    }
+    return KernelExecutorImpl::CheckResultType(out, function_name);
+  }
```
Comment (Crystrix) [inline cpp/src/arrow/compute/exec.cc:970]: I am not sure how to return different types of data, so I just skip the `CheckResultType` for `sum(null)`.

File: cpp/src/arrow/compute/kernels/aggregate_basic_internal.h [hunk 2]
```diff
@@ -164,6 +195,20 @@ struct MeanImpl : public SumImpl<ArrowType, SimdLevel> {
...
 
+struct NullMeanImpl : public NullSumImpl {
+  explicit NullMeanImpl(const ScalarAggregateOptions& options_) : NullSumImpl(options_) {}
+
+  Status Finalize(KernelContext*, Datum* out) override {
+    if ((options.skip_nulls || this->is_empty) && options.min_count == 0) {
+      // Return 0 if the remaining data is empty
+      out->value = std::make_shared<DoubleScalar>(0);
+    } else {
+      out->value = MakeNullScalar(float64());
```
Comment (Crystrix) [inline cpp/src/arrow/compute/kernels/aggregate_basic_internal.h:206]: `mean(null)` is different with `sum(null)` as the returned value should be a `DoubleScalar`

**Comment by Crystrix** [comment]
This PR is rebased with the master branch and updated.
- The MinMax of null type is supported now, so it's removed in this PR.
- The sum of an empty array or `min_count == 0` returns an int64 scalar of 0, otherwise return a null scalar
- The mean of an empty array or `min_count == 0` returns a double scalar of 0, otherwise return a double scalar of null value

I am not sure how to return different types of data, so I skipped the type checking.

@jorisvandenbossche @pitrou @bkietz Could you please help to review?

File: cpp/src/arrow/compute/exec.cc (hunk 1 above)
Comment (lidavidm) [reply cpp/src/arrow/compute/exec.cc:970]: Let's not do this. Below we should return a null scalar of the right type, not a null scalar of the null type.

File: cpp/src/arrow/compute/kernels/aggregate_basic.cc [hunk 3]
```diff
@@ -929,6 +929,7 @@ void RegisterScalarAggregateBasic(FunctionRegistry* registry) {
   AddArrayScalarAggKernels(SumInit, SignedIntTypes(), int64(), func.get());
   AddArrayScalarAggKernels(SumInit, UnsignedIntTypes(), uint64(), func.get());
   AddArrayScalarAggKernels(SumInit, FloatingPointTypes(), float64(), func.get());
+  AddArrayScalarAggKernels(SumInit, {null()}, null(), func.get());
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/aggregate_basic.cc:932]: If we're returning int64 below, this should be declared to return int64.

File: cpp/src/arrow/compute/kernels/aggregate_basic_internal.h [hunk 4]
```diff
@@ -121,6 +121,37 @@ struct SumImpl : public ScalarAggregator {
...
+    this->is_empty &= other.is_empty;
+    return Status::OK();
+  }
+
+  Status Finalize(KernelContext*, Datum* out) override {
+    if ((options.skip_nulls || this->is_empty) && options.min_count == 0) {
+      // Return 0 if the remaining data is empty
+      out->value = std::make_shared<Int64Scalar>(0);
+    } else {
+      out->value = MakeNullScalar(null());
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/aggregate_basic_internal.h:146]:
```suggestion
      out->value = MakeNullScalar(int64());
```

**Comment by Crystrix** [comment]
@lidavidm, the code is updated. Now the sum of a null array always returns an int64 scalar.

**Review by lidavidm — APPROVED** [review]
LGTM, thanks for the update.

**Pull request closed without merging**